### PR TITLE
subscription: Add AttachedSubscription DBus structure

### PR DESCRIPTION
**Pull request by M4rtinK**
The AttachedSubscription DBus structure describes a single subscription
that has been attached to the target system.

A list of AttachedSubscription instances together with the InsightsEnabled property
and SystemPurposeData property describe the system state post registration and subscription.

Note that all the fields are effectively string "blobs" we are getting
from a JSON file returned by RHSM DBus API.

### CODE REVIEW

File: pyanaconda/modules/common/structures/subscription.py [hunk 1]
```diff
@@ -362,3 +362,107 @@ def server_proxy_password(self) -> SecretData:
...
+        """
+        return self._name
+
+    @name.setter
+    def name(self, name: Str):
+        self._name = name
+
+    @property
+    def service_level(self) -> Str:
+        """Service level of the attached subscription.
```
Comment (poncovka) [inline pyanaconda/modules/common/structures/subscription.py:394]: Could you provide an example in the docstring?

File: pyanaconda/modules/common/structures/subscription.py [hunk 2]
```diff
@@ -362,3 +362,107 @@ def server_proxy_password(self) -> SecretData:
...
+        """
+        return self._service_level
+
+    @service_level.setter
+    def service_level(self, service_level: Str):
+        self._service_level = service_level
+
+    @property
+    def sku(self) -> Str:
+        """SKU id of the attached subscription.
```
Comment (poncovka) [inline pyanaconda/modules/common/structures/subscription.py:407]: Same here.

File: pyanaconda/modules/common/structures/subscription.py [hunk 3]
```diff
@@ -362,3 +362,107 @@ def server_proxy_password(self) -> SecretData:
...
+        """
+        return self._sku
+
+    @sku.setter
+    def sku(self, sku: Str):
+        self._sku = sku
+
+    @property
+    def contract(self) -> Str:
+        """Contract identifier.
```
Comment (poncovka) [inline pyanaconda/modules/common/structures/subscription.py:420]: Same here?

File: pyanaconda/modules/common/structures/subscription.py [hunk 4]
```diff
@@ -362,3 +362,107 @@ def server_proxy_password(self) -> SecretData:
...
+        """
+        return self._contract
+
+    @contract.setter
+    def contract(self, contract: Str):
+        self._contract = contract
+
+    @property
+    def start_date(self) -> Str:
+        """Subscription start date.
```
Comment (poncovka) [inline pyanaconda/modules/common/structures/subscription.py:433]: Are we able to specify the format? I know we had some issues with that.

File: tests/nosetests/pyanaconda_tests/module_subscription_tests.py [hunk 5]
```diff
@@ -798,6 +798,57 @@ def set_subscription_request_sensitive_data_keep_test(self):
...
+            "service-level": "really good",
+            "sku": "ABC123456",
+            "contract": "987654321",
+            "start-date": "1 January 1970",
+            "end-date": "19 January 2038",
+            "consumed-entitlement-count": "9001"
+        }
+        # compare the full structure with expected values
+        self.assertEqual(
+            get_native(AttachedSubscription.to_structure(full_request)),
```
Comment (poncovka) [inline tests/nosetests/pyanaconda_tests/module_subscription_tests.py:848]: It would be good to have one test with variants so we can verify the types.

File: pyanaconda/modules/common/structures/subscription.py [hunk 6]
```diff
@@ -362,3 +362,107 @@ def server_proxy_password(self) -> SecretData:
...
+        :rtype: str
+        """
+        return self._end_date
+
+    @end_date.setter
+    def end_date(self, end_date: Str):
+        self._end_date = end_date
+
+    @property
+    def consumed_entitlement_count(self) -> Str:
```
Comment (poncovka) [inline pyanaconda/modules/common/structures/subscription.py:458]: Shouldn't this be an integer?

Comment (M4rtinK) [reply pyanaconda/modules/common/structures/subscription.py:458]: The idea is that we are not really sure - the JSON returned by the RHSM DBus API from which this data is returned has what seems to be a string containing an integer. But given that the format is not well documented and I fear we might one day get "3 entitlements" or something like that.

So it seems prudent to me to treat this data as an opaque string that we do not really interpret to avoid surprises later on.

File: pyanaconda/modules/common/structures/subscription.py (hunk 4 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/structures/subscription.py:433]: The problem is in the JSON from which this is parsed providing dates in an inconsistent manner - for more info see [bug 1793501](https://bugzilla.redhat.com/show_bug.cgi?id=1793501).

The code that does the parsing (coming soon in followup PR) handles these differences and provides a consistent human readable data that is set to the DBus structure and then displayed.

Another issues is that we would like to harmonize the date display between all the tools handling this data (Anaconda, Cockpit RHSM plugin, RHSM CLI, Satellite, etc.) but this is still an ongoing effort and no shared date format has yet been agreed on.

To summarize, the only date format we can specify at the moment for these fields in the DBus structure is "nice human readable date with consistent formatting" - which I will add to the docstring. :)

Later on (possibly before declaring our DBus API stable) we might able able to improve this.

File: pyanaconda/modules/common/structures/subscription.py (hunk 6 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/structures/subscription.py:458]: This is one of the bugs for the JSON output not being documented: [bug 1790987](https://bugzilla.redhat.com/show_bug.cgi?id=1790987)

Comment (poncovka) [reply pyanaconda/modules/common/structures/subscription.py:458]: In that case, could you document the explanation in the docstring of the property?

File: pyanaconda/modules/common/structures/subscription.py (hunk 4 above)
Comment (poncovka) [reply pyanaconda/modules/common/structures/subscription.py:433]: Ok, I would document this in the docstring.

File: pyanaconda/modules/common/structures/subscription.py (hunk 1 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/structures/subscription.py:394]: Added an example.

File: pyanaconda/modules/common/structures/subscription.py (hunk 2 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/structures/subscription.py:407]: Added an example.

File: pyanaconda/modules/common/structures/subscription.py (hunk 3 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/structures/subscription.py:420]: Added.

File: pyanaconda/modules/common/structures/subscription.py (hunk 4 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/structures/subscription.py:443]: Added a docstring explaining what can be expected in the data property & linked to the bug.

File: pyanaconda/modules/common/structures/subscription.py (hunk 6 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/structures/subscription.py]: So I looked deeper into this and in the end it looks rather unlikely we will get something else than an integer in the key we are parsing in the JSON file. Also the "consumed" suffix is added by us at the moment. So I've switched the property to be an `Int`.

File: tests/nosetests/pyanaconda_tests/module_subscription_tests.py (hunk 5 above)
Comment (M4rtinK) [reply tests/nosetests/pyanaconda_tests/module_subscription_tests.py]: Good point, switched the tests to variants. :)

File: pyanaconda/modules/common/structures/subscription.py (hunk 3 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/structures/subscription.py:428]: Added.

File: pyanaconda/modules/common/structures/subscription.py (hunk 2 above)
Comment (M4rtinK) [reply pyanaconda/modules/common/structures/subscription.py:413]: Added.

**Comment by M4rtinK** [comment]
PR updated
- examples added
- entitlement count switched to Int
- improved tests
